src/nlsolve/anderson.jl

```diff
@@ -58,156 +57,148 @@ function (S::NLAnderson{false,<:NLSolverCache})(integrator)
     gs[1] = z₊
 
     mk = min(S.n, iter-1)
-    ztmp = zs[1] .- gs[1]
-    for i in 2:mk+1
-      @. residuals[:,i-1] = gs[i] - zs[i] + ztmp
-    end
-    resqr = qr(residuals[:,1:mk], Val(true))
-    if typeof(ztmp) <: AbstractArray
-      alphas[1:mk] = resqr \ ztmp
-    else
-      alphas[1:mk] = resqr \ [ztmp]
+    @. ztmp = zs[1] - gs[1]
```

Comment:
shouldn't assume mutability when not in place

Reply:
Is it not OK even if `ztmp` is defined as an array in https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/pull/638/files/fe91244da8c13b9f29c7d62d339f253541bb25b9#diff-8118feaffe7fe662c953104e316305afR11?

Reply:
That would work, but why not just keep it a type-matching value so that way SArrays work without a conversion to a vector here?

Reply:
I think the only thing that truly needs to be an array is `residuals`?

Reply:
Managing the case where `residuals` is an SArray` would be a good optimization to do some time in the future though.

Reply:
The main issue I saw with the current implementation are lines such as https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/pull/638/files#diff-8118feaffe7fe662c953104e316305afL63, where it is implicitly assumed that `zs[i]`, `gs[i]` and `ztmp` are one-dimensional arrays (or scalars). The next thing I noticed was that since `residuals` is defined as an Array, the QR decomposition `qr` in https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/pull/638/files#diff-8118feaffe7fe662c953104e316305afL65 could be replaced by `qr!`, as in the in-place version. Similarly, the left division in line https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/pull/638/files#diff-8118feaffe7fe662c953104e316305afL67 and https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/pull/638/files#diff-8118feaffe7fe662c953104e316305afL69 can be replaced by the in-place `ldiv!` as long as `ztmp` is guaranteed to be an Array, since also `alphas` is already defined as an Array. Hence I assumed the easiest way would be to define `ztmp` as an Array and just update it in every iteration.

Reply:
It seems that the standard implementation of `ldiv!` in LinearAlgebra doesn't work with a `ztmp` of type StaticArray due to https://github.com/JuliaArrays/StaticArrays.jl/issues/589. Otherwise it would be possible to use the implementation
```julia
ztmp = zs[1] .- gs[1]
vecztmp = _vec(ztmp)
for i in 2:(mk + 1)
    residuals[:, i - 1] .= _vec(gs[i]) .- _vec(zs[i]) .+ vecztmp
end
resqr = qr!(view(residuals, :, 1:mk), Val(true))
ldiv!(view(alphas, 1:mk), resqr, vecztmp)
for i in 1:mk
    z₊ = @. z₊ + alphas[i] * (gs[i+1] - gs[1])
end
```